### ISSUE TITLE
TCP support in SignalR

### ISSUE THREAD

**Issue by Suvarnnan** [aspnet/SignalR#1291]
ASP.NET Core 2.0 SignalR supports different middle ware options like TCP,AMQP etc.

Can anyone give details about it like a sample where TCP is used as signalr middle ware.

Any good explanation of this concept and illustration with samples are appreciated.

**Comment by anurse** [comment]
> ASP.NET Core 2.0 SignalR supports different middle ware options like TCP,AMQP etc.

We don't currently support TCP or AMQP. Any discussions we've had about supporting that just refer to the fact that our architecture allows us to do that in the future. There is no support for TCP or AMQP in any official build at the moment.

**Comment by Suvarnnan** [comment]
Thanks for your quick response .
Can you inform us when this will be available   ?
We are looking for this information curiously .

**Comment by anurse** [comment]
There is no roadmap for TCP support in the upcoming 2.1 release. The infrastructure is in place but we have no plans to support it yet. We may look at it after 2.1 is released, but we have a enough work just getting to our first major release in 2.1 :).

**Comment by davidfowl** [comment]
Official TCP support will be done in 2.2, we already have all the pieces to make the server work (see https://github.com/aspnet/KestrelHttpServer/issues/1980#issuecomment-377123245)

**Comment by davidfowl** [comment]
We have a TCP sample now:

**Client**

https://github.com/aspnet/SignalR/blob/4ddf8664c0a8da3b14f147c46b971b73a39538e9/samples/ClientSample/HubSample.cs#L31-L40

**Server**

https://github.com/aspnet/SignalR/blob/4ddf8664c0a8da3b14f147c46b971b73a39538e9/samples/SignalRSamples/Program.cs#L35-L38

**Comment by davidfowl** [comment]
https://github.com/aspnet/SignalR/tree/release/2.1/samples/ClientSample/Tcp

**Comment by TrabacchinLuigi** [comment]
I'll try! 
Thanks! you're a good dude! xD

**Comment by NithinKB666** [comment]
> We have a TCP sample now:
> 
> **Client**
> 
> [SignalR/samples/ClientSample/HubSample.cs](https://github.com/aspnet/SignalR/blob/4ddf8664c0a8da3b14f147c46b971b73a39538e9/samples/ClientSample/HubSample.cs#L31-L40)
> 
> Lines 31 to 40 in [4ddf866](/aspnet/SignalR/commit/4ddf8664c0a8da3b14f147c46b971b73a39538e9)
> 
>  var endpoint = new IPEndPoint(IPAddress.Loopback, 9001); 
>  Console.WriteLine("Connecting to {0}", endpoint); 
>  var connection = new HubConnectionBuilder() 
>      .WithEndPoint(endpoint) 
>      .WithLogging(logging => 
>      { 
>          logging.AddConsole(); 
>          logging.SetMinimumLevel(LogLevel.Trace); 
>      }) 
>      .Build(); 
> **Server**
> 
> [SignalR/samples/SignalRSamples/Program.cs](https://github.com/aspnet/SignalR/blob/4ddf8664c0a8da3b14f147c46b971b73a39538e9/samples/SignalRSamples/Program.cs#L35-L38)
> 
> Lines 35 to 38 in [4ddf866](/aspnet/SignalR/commit/4ddf8664c0a8da3b14f147c46b971b73a39538e9)
> 
>  options.Listen(IPAddress.Any, 9001, builder => 
>  { 
>      builder.UseHub<Chat>(); 
>  });

Can I use Streaming with channels with this?

**Comment by anurse** [comment]
Yes, the full set of SignalR features should work over the TcpConnection sample, though you won't have access to an `HttpContext` on the server, because there is no HTTP :).

**Comment by davidfowl** [comment]
See https://docs.microsoft.com/en-us/aspnet/core/signalr/security?view=aspnetcore-2.1#buffer-management for HTTP configuration.  

The TCP limits should be a 1MB by default (Kestrel limits). But you may need to bump it up a bit to account for the SignalR framing.

**Comment by NithinKB666** [comment]
> See https://docs.microsoft.com/en-us/aspnet/core/signalr/security?view=aspnetcore-2.1#buffer-management for HTTP configuration.
> 
> The TCP limits should be a 1MB by default (Kestrel limits). But you may need to bump it up a bit to account for the SignalR framing.

Thanks david,
 I increased the Kestrel limits even though it dint work.
Then In the client TcpConnection.cs file.

private async Task ProcessReceives()
        {
            
                while (true)
                {
                    var buffer = _application.Output.GetMemory(1000000);

                    var bytesReceived = await _receiver.ReceiveAsync(buffer);

                    if (bytesReceived == 0)
                    {
                        // FIN
                        break;
                    }

                    _application.Output.Advance(bytesReceived);

                    var flushTask = _application.Output.FlushAsync();

                    if (!flushTask.IsCompleted)
                    {
                        await flushTask;
                    }

                    var result = flushTask.GetAwaiter().GetResult();
                    if (result.IsCompleted)
                    {
                        break;
                    }
                }
            }
          
 

I added 1000000 in getMemory(), which seems to be working. thanks again.

**Comment by NithinKB666** [comment]
I have a question...
I am planning to use one hub to handle TCP and another to handle http clients. To get the data of 1mb(max), do you think it's better to use aspnet core web api to get this 1mb data or is it fine using signalR to get the data. Is there any performance issue using signalR to get such a big data over web api? At max I will have 10 http clients and 5 TCP clients. Your comment on this is highly appreciated.
